miden-tx/src/host/account_procs.rs

```diff
@@ -1,40 +1,89 @@
 
 // ACCOUNT PROCEDURE INDEX MAP
 // ================================================================================================
 
 /// A map of proc_root |-> proc_index for all known procedures of an account interface.
-pub struct AccountProcedureIndexMap(BTreeMap<Digest, u8>);
+pub struct AccountProcedureIndexMap(BTreeMap<Digest, u16>);
 
 impl AccountProcedureIndexMap {
     /// Returns a new [AccountProcedureIndexMap] instantiated with account procedures present in
     /// the provided advice provider.
-    ///
-    /// This function assumes that the account procedure tree (or a part thereof) is loaded into the
-    /// Merkle store of the provided advice provider.
-    pub fn new<A: AdviceProvider>(account_code_root: Digest, adv_provider: &A) -> Self {
-        // get the Merkle store with the procedure tree from the advice provider
-        let proc_store = adv_provider.get_store_subset([account_code_root].iter());
+    pub fn new<A: AdviceProvider>(
+        account_code_commitment: Digest,
+        adv_provider: &A,
+    ) -> Result<Self, TransactionHostError> {
+        // get the account procedures from the advice_map
+        let proc_data =
+            adv_provider.get_mapped_values(&account_code_commitment).ok_or_else(|| {
+                TransactionHostError::AccountProcedureIndexMapError(
+                    "Failed to read account procedure data from the advice provider".to_string(),
+                )
+            })?;
 
-        // iterate over all possible procedure indexes
         let mut result = BTreeMap::new();
-        for i in 0..AccountCode::MAX_NUM_PROCEDURES {
-            let index = NodeIndex::new(AccountCode::PROCEDURE_TREE_DEPTH, i as u64)
-                .expect("procedure tree index is valid");
-            // if the node at the current index does not exist, skip it and try the next node;this
-            // situation is valid if not all account procedures are loaded into the advice provider
-            if let Ok(proc_root) = proc_store.get_node(account_code_root, index) {
-                // if we got an empty digest, this means we got to the end of the procedure list
-                if proc_root == Digest::default() {
-                    break;
-                }
-                result.insert(proc_root, i as u8);
-            }
+
+        // sanity checks
+
+        // check that there are procedures in the account code
+        if proc_data.is_empty() {
+            return Err(TransactionHostError::AccountProcedureIndexMapError(
+                "The account code does not contain any procedures.".to_string(),
+            ));
         }
-        Self(result)
+
+        // check that the account code does not contain too many procedures
+        if proc_data[0].as_int() > AccountCode::MAX_NUM_PROCEDURES as u64 {
+            return Err(TransactionHostError::AccountProcedureIndexMapError(
+                "The account code contains too many procedures.".to_string(),
+            ));
```

Comment:
nit: I would define a variable for `num_procs` and then use it throughout. This would make the code more readable. For example:
```rust
let num_procs = proc_data[0].as_int() as usize;
if num_procs > AccountCode::MAX_NUM_PROCEDURES {
    return Err(TransactionHostError::AccountProcedureIndexMapError(
        "The account code contains too many procedures.".to_string(),
    ));
}
```

Reply:
Added.